### ISSUE TITLE
index_copy_ costs O(destination) instead of O(payload)

### ISSUE THREAD

**Issue by jvlunteren** [torch-spyre/torch-spyre#4239]
### What happened?

Tensor`.index_copy_(0, idx, src)` on spyre costs time proportional to the destination size, not the number of rows copied. Writing 8 rows into a destination scales linearly with that destination, at a roughly constant implied bandwidth. This is consistent with the op reading and writing the whole destination to place a few rows.

Measured with the per-call launch floor subtracted (`torch.spyre.synchronize()` on both sides of the timed region), 8 rows written in every case:

```
┌─────────────┬─────────────┬───────────────────┐
│ destination │ device time │ implied bandwidth │
├─────────────┼─────────────┼───────────────────┤
│ 1×          │ 1.00×       │ 1.00×             │
├─────────────┼─────────────┼───────────────────┤
│ 4×          │ 4.00×       │ 1.00×             │
├─────────────┼─────────────┼───────────────────┤
│ 16×         │ 16.7×       │ 0.96×             │
└─────────────┴─────────────┴───────────────────┘
```

Holding the destination fixed and varying the payload 16× (1 -> 16 rows) changes the time by under 1%.

### What did you expect to happen?

Cost proportional to the payload: O(len(idx)), not O(dst.numel()). A scatter of 8 rows should not touch 32 MB.

### How can we reproduce it?

```
import os, time
for k, v in (("RANK","0"), ("LOCAL_RANK","0"), ("WORLD_SIZE","1"),
             ("LOCAL_WORLD_SIZE","1"), ("MASTER_ADDR","127.0.0.1"),
             ("MASTER_PORT","29500")):
    os.environ.setdefault(k, v)

import torch
import torch_spyre

KV, D, ITERS = 8, 128, 20

def bench(n_slots, n_tok):
    dst = torch.zeros(n_slots, KV, D, dtype=torch.float16).to("spyre")
    src = torch.randn(n_tok, KV, D, dtype=torch.float16).to("spyre")
    idx = torch.arange(n_tok, dtype=torch.int64).to("spyre")
    fn = torch.compile(lambda d, i, s: d.index_copy_(0, i, s), dynamic=False)
    for _ in range(3):                       # warm compile out of the timing
        fn(dst, idx, src)
    torch.spyre.synchronize()
    t0 = time.perf_counter()
    for _ in range(ITERS):
        fn(dst, idx, src)
    torch.spyre.synchronize()
    return (time.perf_counter() - t0) / ITERS * 1e3

print("payload varies, destination fixed at 4096 slots")
for n_tok in (1, 4, 8, 16):
    print(f"  {n_tok:>3} rows  {bench(4096, n_tok):.4f} ms")

print("destination varies, 8 rows written each time")
for n_slots in (8, 1024, 4096, 16384):       # 8 slots = launch-cost floor
    print(f"  {n_slots:>6} slots  {bench(n_slots, 8):.4f} ms")
```

The first loop should scale with the payload but is flat. The second should be flat but scales with the destination. Subtract the 8-slot row from the others to remove the per-call launch cost, which otherwise masks the scaling at small sizes.

### Any environmental details we need to know?

torch-spyre a96bb86

### Anything else we need to know?

This is the dominant cost in paged-KV decode, where each step scatters a handful of rows into a large cache.

### Relevant log output

```Shell

```

**Comment by tdoublep** [comment]
I think the root cause of this is https://github.com/torch-spyre/torch-spyre/issues/4115 which already has an in-flight fix: https://github.com/torch-spyre/torch-spyre/pull/4163

We need to validate in spyre-inference

**Comment by jvlunteren** [comment]
Retested after #4163 (`da15ede`): fixed for layout-compliant destinations.

Single-commit delta vs parent `cec0ce0`, non-profiler build, 8 rows written, `restickify_` count from generated code:

```
| dest  | `cec0ce0`          | `da15ede`          |
|-------|--------------------|--------------------|
| 2 MB  | 9 copies, 0.178 ms | 0 copies, 0.095 ms |
| 8 MB  | 9 copies, 0.462 ms | 0 copies, 0.126 ms |
| 32 MB | 9 copies, 1.482 ms | 0 copies, 0.095 ms |
```

Destination-size scaling goes from 8.3× to flat (0.99×). The real vLLM KV cache is compliant (`slot_major_kv_layout` puts the slot axis outermost) and shows 0 copies, flat 0.09–0.11 ms from 2-32 MB.

Closing as fixed by #4163.

**Comment by jvlunteren** [comment]
Fixed by #4163
